exercises/practice/robot-simulator/lib/robot_simulator.ex

```diff
@@ -1,10 +1,15 @@
 defmodule RobotSimulator do
+  @type robot() :: any()
+  @type direction() :: :north | :east | :south | :west
+  @type position :: {integer(), integer()}
+
   @doc """
   Create a Robot Simulator given an initial direction and position.
 
   Valid directions are: `:north`, `:east`, `:south`, `:west`
   """
-  @spec create(direction :: atom, position :: {integer, integer}) :: any
+  @spec create(direction, position) ::
+    robot() | {:error, "invalid direction" | "invalid position"}
```

Comment:
I know this is more descriptive but I wouldn't spell out the individual errors, using `String.t()` would be fine. Same for `simulate` below.

Reply:
The point is that without spelling of individual errors it is still underspecified. I need to run tests to understand what task expects out of me, as the tests check for **fixed** strings. If it would simply check for `{:error, _}` then it could specify `{:error, String.t()}`. The whole PR is just to state this strings there.

Reply:
I'll take a closer look in a day or two because I'm pretty busy but I just want to say:
>  I need to run tests to understand what task expects out of me

This is the idea behind every single practice exercise on Exercism

Reply:
I think spelling out the individual error string codes is not typical of Elixir code that I have reviewed in the past.  I would agree with @jiegillet's suggestion to keep it as `String.t()`.


Agree with @angelikatyborska, this is one of the limitations of using a central repository for the problem descriptions that all language tracks suffer from. -- the solution requirements for a particular language are not clearly always spelled out in the problem description.  However, Exercism also encourages TDD in practice tests, so you are reading the tests and encountering those error strings as you progress through the test cases.

Reply:
> I think spelling out the individual error string codes is not typical of Elixir code that I have reviewed in the past. I would agree with @jiegillet's suggestion to keep it as `String.t()`.

Using strings as error values is not common thing in Elixir. And when you use atoms then you often have list of them available somewhere.

Additionally Elixir approach is to return `{:ok, term()} | {:error, term()}`, not `term() | {:error, term()}`. So if the specs for the task would say `{:ok, term()} | {:error, term()}` it would be more clear that the task is to write code in a way that it should handle the errors in meaningful way. With current description I just assumed that the caller will always will be well-behaved and in case of spec mismatch it should fail. Because type explicitly says that the position will be tuple of 2 integers, I just assume that it will never mismatch the specs, as that is counterintuitive. The [tests does that](https://github.com/exercism/elixir/blob/40227b8141577eadcbda120f10128abbf6725b7f/exercises/practice/robot-simulator/test/robot_simulator_test.exs#L45) and it is confusing and frustrating.

---

IMHO the task description should be enough for writing the proper solution on the first try for experienced developer. That is frustrating if I need to look around for external sources of the information if I want to finish a task.

Reply:
I'm not disagreeing with your suggestion to add the `direction()` type to give context to the atom. Sure, it's a good change.

With regard to your addendum, sure that is an approach to use `{:ok, term()} | {:error, term()}`, and I would be open to that option as well -- but that's not the change you proposed in your PR so I'm sorry if I misunderstood your intention. If @angelikatyborska and/or @jiegillet are open to that change, let's do it if it makes it better and eases frustration.  Likely you are not the only one experiencing this.

Reply:
So my suggestion is one of:

- remove the `spec`s and force user to read tests, as current `spec` is misleading
- remove tests that checks for error values, as these are against the `spec`